Add entity field to a graphql fragment

rastiQ opened the pull request.

jozef-slezak commented on packages/graphql-lowcode/__tests__/testScenarios.ts:
```diff
@@ -0,0 +1,126 @@
+ */
+
+export const diffTypeQuery = 
+`{
```
please, add few more test scenarios please with test data that also contain query like this:
```
query x {
  ...fragmentName
}
fragment fragmentName on diffEntity {
    field1
}`
```

we will ensure that it will work also for real worlds scenarios

jozef-slezak commented on packages/graphql-lowcode/src/modify/addFragmentFields.ts:
```diff
@@ -0,0 +1,43 @@
+  const indentSize = format === '\n' ? 2 : 1
+  let indent = 0
+
+  for (const char of graphqlQuery.substr(firstBracket)) {
```
Please move the code that guess indentation in a specific function. This will help another coder to read the code more intuitively.

jozef-slezak commented:
@rastiQ, nice work! Please check the review comments. We are close to merging this pull request.

rastiQ commented:
- Added more real world test scenarios
- Refactored indentation, now it gets whitespaces from existing fields in fragments

In test scenario below I left appending the new field on new line (to achieve git diff highlight only on the new line)

```
query x {
  ...fragmentName
}
fragment fragmentName on diffEntity {
    field1 field2 field3
}`
```

jozef-slezak commented on packages/graphql-lowcode/src/modify/addFragmentFields.ts:
```diff
@@ -0,0 +1,72 @@
+        const endIndex = lastFragmentField?.loc?.end
+
+        //if is the first iteration get first get first fragment field for formatting
+        if(firstFragmentFieldIndex == -1) firstFragmentFieldIndex = startIndex ? startIndex : -1
```
This `if` can be simplified:
`if(firstFragmentFieldIndex == -1 && startIndex) firstFragmentFieldIndex = startIndex`

jozef-slezak commented on packages/graphql-lowcode/src/modify/addFragmentFields.ts:
```diff
@@ -0,0 +1,72 @@
+}
+
+export function fragmentContainsField(fragmentQuery: string, fieldName: string) {
+  if (fragmentQuery.indexOf(fieldName) >= 0) return true
```
this can cause bugs: it would return true even for prefixes...

jozef-slezak commented on packages/graphql-lowcode/src/modify/addFragmentFields.ts:
```diff
@@ -0,0 +1,72 @@
+  if (!graphqlQuery.indexOf('fragment')) return graphqlQuery
+
+  const ast = parse(graphqlQuery)
+  let finalQuery = graphqlQuery
```
move this variable close to the place where it is used for the first time
initialize with an empty string (see the reason in comment below)

jozef-slezak commented on packages/graphql-lowcode/src/modify/addFragmentFields.ts:
```diff
@@ -0,0 +1,72 @@
+
+  let offset = 0
+  lastFragmentFieldIndexes.forEach(index => {
+    finalQuery = finalQuery.slice(0, index + offset) + `${format}${indentString}${options.property}` + finalQuery.slice(index + offset)
```
it is easier to maintain the code like this:
 - `originalQuery.substring()` rather then `finalQuery.slice()` because the offsets are relative to original query (it is easier...)
 - just append finalQuery with new original query substring

jozef-slezak commented on packages/graphql-lowcode/src/modify/addFragmentFields.ts:
```diff
@@ -0,0 +1,72 @@
+  return finalQuery
+}
+
+export function getIndentationNew(graphqlQuery: string, fragmentStartIndex: number) {
```
let's rename to `existingFieldIdentation()`

jozef-slezak commented on packages/graphql-lowcode/src/modify/addFragmentFields.ts:
```diff
@@ -0,0 +1,72 @@
+    }
+  })
+
+  const { indent, indentString, format } = getIndentationNew(graphqlQuery, firstFragmentFieldIndex)
```
Visit the AST twice please because you do two things: a/ guess identation, b/ find offsets where the `property` is supposed to be added.  AST visitor is a technical detail (very much like for loop). Move the visitor to a helper function (eg. `firstFieldIdentation()`, ``)  We preferer easier to read/maintainable code rather then quicklier.

jozef-slezak commented on packages/graphql-lowcode/src/modify/addFragmentFields.ts:
```diff
@@ -0,0 +1,72 @@
+
+  const { indent, indentString, format } = getIndentationNew(graphqlQuery, firstFragmentFieldIndex)
+
+  let offset = 0
```
move this code to a local helper function please: `patchQueryFields()`

jozef-slezak commented on packages/graphql-lowcode/src/modify/addFragmentFields.ts:
```diff
@@ -0,0 +1,72 @@
+  const ast = parse(graphqlQuery)
+  let finalQuery = graphqlQuery
+
+  let lastFragmentFieldIndexes: number[] = []
```
this should be a `const` returned by `findQueryOffsetsForField()`

jozef-slezak commented on packages/graphql-lowcode/src/modify/addFragmentFields.ts:
```diff
@@ -0,0 +1,72 @@
+  return { indent, indentString, format}
+}
+
+export function fragmentContainsField(fragmentQuery: string, fieldName: string) {
```
this should be just a local function (not exported)

rastiQ commented:
- Readability (code divided to more functions)
- Different approach in indentation
- Potentional bug fix (checking if fragment already does not contain new field)
- Visiting AST twice (once for indentation and once for offset)
